### ISSUE TITLE
Check Atom lookup

### ISSUE THREAD

**Issue by uli42** [ArcticaProject/nx-libs#357]
While hunting down some valgrind findings I extended `NameForAtom()` in `dix/atom.c` by an fprint:

```
char *
NameForAtom(Atom atom)
{
    NodePtr node;
    if (atom > lastAtom)
      return NULL;
    if ((node = nodeTable[atom]) == (NodePtr)NULL)
      return NULL;
    fprintf(stderr, "NameForAtom [%d] is [%s]\n", atom, node->string);
    return node->string;
}
```

I then got a print for every successful run of that function. The interesting part: Almost all Lookup appeared twice in a row, with `TARGETS` being the only exception. Example output:

```
...
NameForAtom [348] is [_NET_CLIENT_LIST]
NameForAtom [348] is [_NET_CLIENT_LIST]
NameForAtom [349] is [_NET_CLIENT_LIST_STACKING]
NameForAtom [349] is [_NET_CLIENT_LIST_STACKING]
NameForAtom [254] is [SM_CLIENT_ID]
NameForAtom [254] is [SM_CLIENT_ID]
NameForAtom [316] is [_NET_ACTIVE_WINDOW]
NameForAtom [316] is [_NET_ACTIVE_WINDOW]
NameForAtom [75] is [TARGETS]
NameForAtom [348] is [_NET_CLIENT_LIST]
NameForAtom [348] is [_NET_CLIENT_LIST]
NameForAtom [349] is [_NET_CLIENT_LIST_STACKING]
NameForAtom [349] is [_NET_CLIENT_LIST_STACKING]
NameForAtom [318] is [_NET_FRAME_EXTENTS]
NameForAtom [318] is [_NET_FRAME_EXTENTS]
NameForAtom [318] is [_NET_FRAME_EXTENTS]
NameForAtom [318] is [_NET_FRAME_EXTENTS]
NameForAtom [373] is [_NET_WM_ALLOWED_ACTIONS]
NameForAtom [373] is [_NET_WM_ALLOWED_ACTIONS]
...
``` 

Looks like there's either a bug somewhere or we have potential for optimization...

**Comment by uli42** [comment]
Reason identified:

`Agent.h` defines `validateString` like this:
```
#define validateString(string) ((string) ? (string) : "(null)")
```

And `NameForAtom` is used like this almost everywhere:
```
validateString(NameForAtom(atom))
```

So the macro expansion will result in two calls of `NameForAtom`.

I suggest making `validateString` a function. Any ideas where to put that function? It would perfectly fit into Utils.c but we only have Utils.h. And introducing a new fiel for just one function is a bit overhead, isn't it?

**Comment by Ionic** [comment]
```C
static inline char* validateString (char *string) {
    char *res = string;
    if (res) {
        return res;
    }
    else {
        return "(null)";
    }
}
```

in `Utils.h`?

**Comment by sunweaver** [comment]
Utils.h sounds good...
Mike

On Sat Feb 25 12:16:58 2017 GMT+0100, Mihai Moldovan wrote:
> static inline validateString (const char *string) {
>     char *res = string;
>     if (res) {
>         return res;
>     }
>     else {
>         return "(null)";
>     }
> }
> in Utils.h? 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or mute the thread.￼
>

-- 
Sent from my Fairphone (running Sailfish OS)...

**Comment by uli42** [comment]
Why did you add the `res` var? In a function we can simply use/return `string`, don't we? The problem we are aiming to solve is solely caused by double macro replacement.

**Comment by Ionic** [comment]
I was thinking about "caching" and got carried away. You're right, naturally, the variable is completely redundant.

**Comment by uli42** [comment]
Closed by #359
